### PR TITLE
fix(ngUpgrade): downgrade projection

### PR DESCRIPTION
**Pull request by petebacondarwin**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[x] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

This is just a re-jigging of #9538 - @youdz please take a look

### CODE REVIEW

**Comment by petebacondarwin** [comment]
@youdz - I made a few API changes. I hope these fit with you.
In particular, I made the `declareNg1Module` private for now - it is not really needed publicly and it is harder to make things private later than the other way around.
I also renamed some stuff...
* `PromiseCompleter` -> `Deferred`
* `initNg1ModuleForTests` -> `registerNg1ModuleForTests`
* the "spyIng" stuff becomes: `CapturedContentSelectors` and `RuntimeCompilerSpyFactory`

I broke the commits up into smaller pieces. It might be worth reviewing commit by commit.
The two main commits I kept with you as the author. Please check that you are happy with that.

**Comment by mhevery** [comment]
This deals with compiler, and so I think @tbosch should have a look.

**Comment by youdz** [comment]
Awesome @petebacondarwin, I can't take a look right now but I'll review first thing in the morning tomorrow. I'm on the East Coast, so hopefully I'll be done before you even start working tomorrow.

**Comment by petebacondarwin** [comment]
Except I'm in UK :-)

File: tools/public_api_guard/upgrade/index.d.ts [hunk 1]
```diff
@@ -1,3 +1,6 @@
+/** @experimental */
+export declare const platformUpgrade: (extraProviders?: Provider[]) => PlatformRef;
```
Comment (petebacondarwin) [inline tools/public_api_guard/upgrade/index.d.ts]: @misko Is there some way so that this doesn't have to be in the public API? I think the only reason we need this here is that we want to access it in our unit tests.

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 2]
```diff
@@ -505,58 +398,274 @@ export class UpgradeAdapter {
...
         }
+        return {
+          post: (scope: angular.IScope, element: angular.IAugmentedJQuery,
+                 attrs: angular.IAttributes, parentInjector: Injector | ParentInjectorPromise,
+                 transclude: angular.ITranscludeFunction): void => {
+            let id = idPrefix + (idCount++);
+            (<any>element[0]).id = id;
+
+            let injectorPromise = new ParentInjectorPromise();
+            element.data(controllerKey(NG2_INJECTOR), injectorPromise);
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: It just occurred to me that we shouldn't be storing a promise under the NG2_INJECTOR key since people might rely upon this being the actual ng2 injector...

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 3]
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
...
+      callback(this.injector);
+    } else {
+      this.callbacks.push(callback);
+    }
+  }
+
+  resolve(injector: Injector) {
+    this.injector = injector;
+
+    // reset the element data to poin to the real injector
```
Comment (gkalpak) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: _poin_ -->  _point_

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 4]
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
...
+  ];
   function directiveFactory(
-      ng1Injector: angular.IInjectorService, componentFactoryRefMap: ComponentFactoryRefMap,
+      ng1Injector: angular.IInjectorService, ng1Compile: angular.ICompileService,
+      capturedContentSelectors: CapturedContentSelectors,
+      componentFactoryRefMap: ComponentFactoryRefMap,
       parse: angular.IParseService): angular.IDirective {
-    var idCount = 0;
+    let idCount = 0;
+    let dashSelector = info.selector.replace(/([A-Z])/g, char => '-' + char.toLowerCase());
```
Comment (gkalpak) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: The parenthesis around `[A-Z]` is unnecessary.

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 5]
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
...
+      compile: (templateElement: angular.IAugmentedJQuery, templateAttributes: angular.IAttributes,
+                transclude: angular.ITranscludeFunction) => {
+        // We might have compile the contents lazily, because this might have been triggered by the
+        // UpgradeNg1ComponentAdapterBuilder, when the ng2 templates have not been compiled yet
+        let linkFns: angular.ILinkFn[];
+        if (capturedContentSelectors.get(dashSelector)) {
+          linkFns = compileProjectedNodes(templateElement);
         }
+        return {
+          post: (scope: angular.IScope, element: angular.IAugmentedJQuery,
```
Comment (gkalpak) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: Nit: You can return the post-link function directly (without the `{post: ...}` object).

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 6]
```diff
@@ -272,6 +272,59 @@ export class UpgradeAdapter {
...
+   * @param modules any Angular 1 modules that the upgrade module should depend upon
+   * @returns an {@link UpgradeAdapterRef}, which lets you register a `ready()` callback to
+   * run assertions once the Angular 2+ components are ready to test through Angular 1.
+   */
+  registerForNg1Tests(modules?: any[]): UpgradeAdapterRef {
+    const windowNgMock = (window as any)['angular'].mock;
+    if (!isPresent(windowNgMock) || !isPresent(windowNgMock.module)) {
+      throw new Error('Failed to find \'angular.mock.module\'.');
+    }
+    this.declareNg1Module(modules);
```
Comment (youdz) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: If I `registerForNg1Tests()`  in a `beforeEach()`, I'd end up re-declaring the upgrader's Angular 1 module for each test. Isn't it better to declare the module once, then call this to only re-run the configuration code every time, not the whole declaration?

If that's the case, then we do need `declareNg1Module()` to be part of the public API.

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 7]
```diff
@@ -272,6 +272,59 @@ export class UpgradeAdapter {
...
+    windowNgMock.module(this.ng1Module.name);
+    const upgrade = new UpgradeAdapterRef();
+    this.ng2BootstrapDeferred.promise.then(
+        () => { (<any>upgrade)._bootstrapDone(this.moduleRef, this.moduleRef); }, onError);
```
Comment (youdz) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: `(<any>upgrade)._bootstrapDone(this.moduleRef, this.ng1Injector)`

**Review by youdz — COMMENTED** [review]
Thanks for the clean-up, especially the comments. When I saw that the currently existing platforms were not commented at all, I figured I'd just follow this. 
But clearly too many comments would have been better than not enough. :-P

File: modules/@angular/upgrade/src/platform_upgrade.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {CompilerFactory, PlatformRef, Provider, createPlatformFactory} from '@angular/core';
```
Comment (youdz) [inline modules/@angular/upgrade/src/platform_upgrade.ts]: Actually it is, the exported `platformUpgrade` is a PlatformRef, so we need to import the type to be able to export it, even if we don't write it explicitly.

**Comment by youdz** [comment]
@googlebot: I'm okay with my commits being contributed to Google.

It would be a pretty bad joke otherwise.

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 9]
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
...
+    const componentFactoryRefMap: ComponentFactoryRefMap = {};
+    const upgradeAdapter = this;
+    const ng1Module = this.ng1Module = angular.module(this.idPrefix, modules);
+    const platformRef = platformUpgrade();
+
+    this.ngZone = new NgZone({enableLongStackTrace: Zone.hasOwnProperty('longStackTraceZoneSpec')});
+    this.ng2BootstrapDeferred = new Deferred();
+    ng1Module.factory(NG2_INJECTOR, () => this.moduleRef.injector.get(Injector))
+        .constant(NG2_ZONE, this.ngZone)
+        .constant(NG2_COMPONENT_FACTORY_REF_MAP, componentFactoryRefMap)
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: I wanted for them to be available to providers...

File: modules/@angular/upgrade/src/upgrade_adapter.ts (hunk 5 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: Yes but I have a feeling that it was done this way for a reason... so we are copying what was already there.

File: modules/@angular/upgrade/src/runtime_compiler_spy.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {RuntimeCompilerFactory} from '@angular/compiler';
+import {COMPILER_OPTIONS, Compiler, CompilerOptions, Inject, Injectable} from '@angular/core';
+import {isPresent} from './facade/lang';
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/runtime_compiler_spy.ts]: Well we have brought the facade folder into the upgrade project for AoT so it is there.
The real question is whether we actually need to use `!isPresent(...)` at all rather than just `!...`

File: modules/@angular/upgrade/src/upgrade_adapter.ts (hunk 6 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: I thought about this for a while but I think the following two points are valid:
* I don't believe that re-declaring is a big performance cost, we do it all the time in our Angular 1 tests. But I would be happy to back down on that if you show me the numbers.
* the trouble is that there is a lot of state being held in the adapter, such as `this.ngZone` and `this.ng2BootstrapDeferred` which really needs reinitializing on each test. We could do a big refactor to make `declareNg1Module` "stateless" but as it is right now I think it is safer to call it every time.

Comment (youdz) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: Sounds good to me, let's stick to redeclaring then. Better safe than fast.

File: modules/@angular/upgrade/src/runtime_compiler_spy.ts (hunk 10 above)
Comment (mhevery) [inline modules/@angular/upgrade/src/runtime_compiler_spy.ts]: Yes please remove all facades as we are actively working on deleting them. Let's not create new ones.

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 11]
```diff
@@ -7,24 +7,26 @@
  */
 
 import {Compiler, CompilerOptions, ComponentFactory, Injector, NgModule, NgModuleRef, NgZone, Provider, Testability, Type} from '@angular/core';
-import {platformBrowserDynamic} from '@angular/platform-browser-dynamic';
 
 import * as angular from './angular_js';
-import {NG1_COMPILE, NG1_INJECTOR, NG1_PARSE, NG1_ROOT_SCOPE, NG1_TESTABILITY, NG2_COMPILER, NG2_COMPONENT_FACTORY_REF_MAP, NG2_INJECTOR, NG2_ZONE, REQUIRE_INJECTOR} from './constants';
+import {NG1_COMPILE, NG1_INJECTOR, NG1_PARSE, NG1_ROOT_SCOPE, NG1_TESTABILITY, NG2_CAPTURED_CONTENT_SELECTORS, NG2_COMPILER, NG2_COMPONENT_FACTORY_REF_MAP, NG2_INJECTOR, NG2_ZONE, REQUIRE_INJECTOR} from './constants';
 import {DowngradeNg2ComponentAdapter} from './downgrade_ng2_adapter';
+import {isPresent} from './facade/lang';
```
Comment (mhevery) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: remove here as well

File: modules/@angular/upgrade/src/util.ts [hunk 12]
```diff
@@ -5,12 +5,8 @@
...
  * found in the LICENSE file at https://angular.io/license
  */
-
-
-export function stringify(obj: any): string {
-  if (typeof obj == 'function') return obj.name || obj.toString();
-  return '' + obj;
-}
+import {CssSelector, SelectorMatcher, createElementCssSelector} from '@angular/compiler';
+import {isPresent} from './facade/lang';
```
Comment (mhevery) [inline modules/@angular/upgrade/src/util.ts]: remove

**Comment by mhevery** [comment]
My only comment is to remove all facades, LGTM otherwise

**Comment by petebacondarwin** [comment]
CI is green and I have rebased on master.
@mhevery happy to merge?

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 13]
```diff
@@ -269,17 +272,70 @@ export class UpgradeAdapter {
...
+   * const upgradeAdapter = new UpgradeAdapter();
+   *
+   * // configure the adapter with upgrade/downgrade components and services
+   * upgradeAdapter.downgradeNg2Component(MyComponent);
+   *
+   * let upgradeAdapterRef: UpgradeAdapterRef;
+   * let $compile, $rootScope;
+   *
+   * beforeEach(() => {
+   *   upgradeAdapterRef = upgradeAdapter.registerForNg1Tests(['heroApp'])
```
Comment (gkalpak) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: Now that you mention it, does it mean that you can use it like:

```js
beforeEach(upgradeAdapter.registerForNg1Tests(['heroApp']))
// Similar to:
// beforeEach(module('heroApp'))
```

If not, then the above comment can be misleading :smiley:

File: modules/@angular/upgrade/src/upgrade_adapter.ts (hunk 9 above)
Comment (gkalpak) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: That's a noble goal indeed :smiley: It comes at the cost of making them "un-decorate-able", of course. No idea which usecase is more useful/common, but I trust you have thought this through.

File: modules/@angular/upgrade/src/platform_upgrade.ts (hunk 8 above)
Comment (gkalpak) [inline modules/@angular/upgrade/src/platform_upgrade.ts]: I didn't understand what you just said, but I am sure you known what you are talking about :smiley:

File: modules/@angular/upgrade/src/upgrade_adapter.ts (hunk 7 above)
Comment (gkalpak) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: Doesn't this need fixing too. BTW, I don't think `this.ng1Injector` works here, because `this.ng1Injector` is set inside `_bootstrapDone`. But now the `ng1Injector` is available as the resolution value for the `ng2BootstrapDeferred.promise`. So, this can be changed to:

```ts
this.ng2BootstrapDeferred.promise.then(
          ng1Injector => { (<any>upgrade)._bootstrapDone(this.moduleRef, ng1Injector); }, onError);
```

(But why isn't any test failing? :confused)

File: modules/@angular/upgrade/src/upgrade_adapter.ts (hunk 13 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: Oh you spoilsport! OK, we need to reword

File: modules/@angular/upgrade/src/upgrade_adapter.ts (hunk 9 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: Undecoratable true - I can't think of a use case for that. You can still override them...

File: modules/@angular/upgrade/src/platform_upgrade.ts (hunk 8 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/platform_upgrade.ts]: Ahah! Now I look again, I agree with @gkalpak - we don't actually use this import

**Comment by youdz** [comment]
@petebacondarwin, @mhevery: Any update on this?

I'd be happy to help, but I'm not sure what to do now that Pete moved to his branch.

**Comment by petebacondarwin** [comment]
@youdz - I'm trying to squeeze it into 2.2.0 which is imminent. If not it might have to wait for 2.3.0

**Comment by petebacondarwin** [comment]
Looks like 2.2.0 is already at RC and so this can't land there. We will get it into 2.3.0 which will be out in a month

**Comment by youdz** [comment]
Hi there, so now that 2.2.0 is out, what can I do to make sure this makes it into 2.3.0? We're trying to align some our timelines based on this, so we really don't want to miss it.

@petebacondarwin: Do you mind if I help by pushing short fixes to your branch, for instance?

**Comment by petebacondarwin** [comment]
I'll rebase and ping @mhevery for a review. 
What small fixes did you have in mind?

**Comment by youdz** [comment]
The build failed because of golden files differences, so I was thinking of that kind of fixes. One or two lines in golden files, maybe one or two if the linter failed, things like that. Nothing too specific for now, I was just asking in general.

**Comment by petebacondarwin** [comment]
It seems that this PR has become stale as the compiler has changed :-(
I'll try a rebase/reworking on Monday...

@tbosch you had some suggestion of how to do this without monkeypatching the compiler?

File: modules/@angular/upgrade/src/platform_upgrade.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    [CapturedContentSelectors, {provide: CompilerFactory, useClass: RuntimeCompilerSpyFactory}];
+
+/**
+ * A custom platform for the UpgradeAdapter, which includes additional helper services
+ * such as the `CapturedContentSelectors` collection and a monkeypatched version of the
+ * `CompilerFactory`.
+ *
+ * @experimental
+ */
+export const platformUpgrade =
```
Comment (tbosch) [inline modules/@angular/upgrade/src/platform_upgrade.ts]: This is only for patching the compiler, right? If we didn't need to patch the compiler, and the compiler had a method that can just give you the content selectors, this would not be needed?

File: modules/@angular/upgrade/src/runtime_compiler_spy.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {RuntimeCompilerFactory} from '@angular/compiler';
```
Comment (tbosch) [inline modules/@angular/upgrade/src/runtime_compiler_spy.ts]: Let's change the `RuntimeCompiler` to have a method `getNgContentSelectors(component: Type<any>)` that uses a cache that is filled during compilation...

Then we don't need this hack...

Comment (tbosch) [inline modules/@angular/upgrade/src/runtime_compiler_spy.ts]: See https://github.com/angular/angular/blob/master/modules/@angular/compiler/src/jit/compiler.ts#L158, where we have the `CompileDirectiveMetadata` which contains the `CompileTemplateMetadata` which contains the `ngContentSelectors`.

Comment (tbosch) [inline modules/@angular/upgrade/src/runtime_compiler_spy.ts]: So you could add this method to the `RuntimeCompiler`:
```
getNgContentSelectors(component: Type<any>): string[] {
  const template = this._compiledTemplateCache.get(component);
  if (!template) {
    throw new Error(`The component ${stringify(component)} is not yet compiled!`);
  }
  return template.compMeta.template.ngContentSelectors
}
```

File: modules/@angular/upgrade/src/platform_upgrade.ts (hunk 14 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/platform_upgrade.ts]: Yes that is right.

File: modules/@angular/upgrade/src/runtime_compiler_spy.ts (hunk 15 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/runtime_compiler_spy.ts]: OK. I will work on that tomorrow. Thanks!

**Comment by jaffoneh** [comment]
hey @petebacondarwin: since this is a change we're depending on moving forward, I was wondering if there is an estimate of when will this make it to Angular 2. I see 2.2.4 was released today and I was wondering if this is going to make into 2.2.5 or if there is a way we could provide any help to accelerate it? Thanks again for your help! cc @youdz

**Comment by petebacondarwin** [comment]
@jaffoneh - this should be ready now. I had to do a bit of refactoring so we were not monkeypatching the compiler.

@tbosch - can you please review the compiler changes?

File: modules/@angular/core/src/linker/compiler.ts [hunk 16]
```diff
@@ -82,6 +82,8 @@ export class Compiler {
     throw _throwError();
   }
 
+  getNgContentSelectors(component: Type<any>): string[] { throw _throwError(); }
```
Comment (tbosch) [inline modules/@angular/core/src/linker/compiler.ts]: Could you add a comment describing the purpose and when this can be called? (e.g. that the component has to be compiled...)

File: modules/@angular/upgrade/src/util.ts [hunk 17]
```diff
@@ -27,3 +22,59 @@ export function onError(e: any) {
...
+    let attrLen = attributes.length;
+    asArray = new Array(attrLen);
+    for (let i = 0; i < attrLen; i++) {
+      asArray[i] = [attributes[i].nodeName, attributes[i].nodeValue];
+    }
+  }
+  return asArray || [];
+}
+
+export function sortProjectableNodes(ngContentSelectors: string[], childNodes: Node[]): Node[][] {
```
Comment (tbosch) [inline modules/@angular/upgrade/src/util.ts]: I think this function deserves some unit tests for all the corner cases (e.g. combination of wildcard and non wildcard ng-content, ...).

File: modules/@angular/upgrade/src/util.ts [hunk 18]
```diff
@@ -27,3 +22,59 @@ export function onError(e: any) {
...
+    }
+    if (ngContentIndices.length > 0) {
+      projectableNodes[ngContentIndices[0]].push(node);
+    }
+  }
+  return projectableNodes;
+}
+
+
+export class Deferred<R> {
```
Comment (tbosch) [inline modules/@angular/upgrade/src/util.ts]: This class seems to be unused...

File: tools/public_api_guard/upgrade/index.d.ts [hunk 19]
```diff
@@ -4,6 +4,7 @@ export declare class UpgradeAdapter {
     bootstrap(element: Element, modules?: any[], config?: angular.IAngularBootstrapConfig): UpgradeAdapterRef;
     downgradeNg2Component(type: Type<any>): Function;
     downgradeNg2Provider(token: any): Function;
+    registerForNg1Tests(modules?: string[]): UpgradeAdapterRef;
```
Comment (tbosch) [inline tools/public_api_guard/upgrade/index.d.ts]: Why is this line added? There is no code that refers to it in this PR...

Comment (tbosch) [inline tools/public_api_guard/upgrade/index.d.ts]: Ah, never mind. Github did not expand all files...

**Comment by tbosch** [comment]
LGTM for everything except modules/@angular/upgrade/src/upgrade_adapter.ts (I don't know the details in there...)

File: modules/@angular/core/src/linker/compiler.ts (hunk 16 above)
Comment (petebacondarwin) [inline modules/@angular/core/src/linker/compiler.ts]: will do!

File: modules/@angular/upgrade/src/util.ts (hunk 17 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/util.ts]: OK

Comment (petebacondarwin) [inline modules/@angular/upgrade/src/util.ts]: done

File: modules/@angular/core/src/linker/compiler.ts (hunk 16 above)
Comment (petebacondarwin) [inline modules/@angular/core/src/linker/compiler.ts]: done

File: modules/@angular/upgrade/test/upgrade_spec.ts [hunk 20]
```diff
@@ -178,7 +178,7 @@ export function main() {
            adapter.bootstrap(element, ['ng1']).ready((ref) => {
              expect(document.body.textContent).toEqual('1A;2A;ng1a;2B;ng1b;2C;1C;');
              // https://github.com/angular/angular.js/issues/12983
-             expect(log).toEqual(['1A', '1B', '1C', '2A', '2B', '2C', 'ng1a', 'ng1b']);
+             expect(log).toEqual(['1A', '1C', '2A', '2B', '2C', 'ng1a', 'ng1b']);
```
Comment (gkalpak) [inline modules/@angular/upgrade/test/upgrade_spec.ts]: Why this change?

File: modules/@angular/upgrade/test/util_spec.ts [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    function html(nodes: Node[]) {
+      const element = document.createElement('div');
+      nodes.forEach(node => element.appendChild(node));
+      return element.innerHTML.trim();
+    }
+
+    describe('sortProjectableNodes', () => {
+      it('should should return an array of node collections for each selector', () => {
```
Comment (gkalpak) [inline modules/@angular/upgrade/test/util_spec.ts]: _should should_ --> _should_

**Review by gkalpak — COMMENTED** [review]
Basically LGTM.
I think there should be some tests for `registerForNg1Tests()`, though.

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 22]
```diff
@@ -271,6 +270,60 @@ export class UpgradeAdapter {
...
+   * });
+   *
+   * ```
+   *
+   * @param modules any Angular 1 modules that the upgrade module should depend upon
+   * @returns an {@link UpgradeAdapterRef}, which lets you register a `ready()` callback to
+   * run assertions once the Angular 2+ components are ready to test through Angular 1.
+   */
+  registerForNg1Tests(modules?: string[]): UpgradeAdapterRef {
+    const windowNgMock = (window as any)['angular'].mock;
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: yes - since the "official" type for `window` does not include index access.

File: modules/@angular/upgrade/src/upgrade_adapter.ts [hunk 23]
```diff
@@ -269,17 +270,71 @@ export class UpgradeAdapter {
...
   /**
-   * Bootstrap a hybrid AngularJS v1 / Angular v2 application.
+   * Registers the adapter's Angular 1 upgrade module for unit testing in Angular 1.
+   * Use this instead of `angular.mock.module()` to load the upgrade module into
+   * the Angular 1 testing injector.
+   *
+   * ### Example
+   *
+   * ```
+   * const upgradeAdapter = new UpgradeAdapter();
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/upgrade_adapter.ts]: OK

**Comment by vicb** [comment]
Fails in G3
`/modules/@angular/upgrade/src/util.ts(8,70): error TS2307: Cannot find module '@angular/compiler'`

**Comment by alexeagle** [comment]
@petebacondarwin we get that error because upgrade cannot depend on compiler. (We have the static-only upgrade)

I think we need a util-static.ts which is included in our library, and a util.ts which is not...

**Comment by petebacondarwin** [comment]
Thanks for looking at this @alexeagle.

So I can put the bits that rely upon the JIT compiler into a separate module, which is not imported by any of the `upgrade/static` modules. Would that solve the problem?

**Comment by alexeagle** [comment]
It doesn't have to be a separate module, we just have the following code that the caretaker might need to modify when you change things:
@vicb you should probably try to fix this yourself since the PR comes from non-googler and it's our own internal build that needs to see the right file subsets....

```
UPGRADE_STATIC_SRCS = [
    "rc/modules/@angular/upgrade/src/aot/**/*.ts",
    "rc/modules/@angular/upgrade/src/facade/**/*.ts",
    "rc/modules/@angular/upgrade/src/angular_js.ts",
    "rc/modules/@angular/upgrade/src/util.ts",
    "rc/modules/@angular/upgrade/static.ts",
]

ng_module(
    name = "upgrade",
    srcs = glob(UPGRADE_STATIC_SRCS),
    deps = [
        ":angular2_es6_subset",
        ":core",
        ":platform_browser",
        "//third_party/javascript/node_modules/rxjs:canary",
    ],
)

# ngUpgrade that depends on compiler
ng_module(
    name = "upgrade_dynamic",
    srcs = glob(
        ["rc/modules/@angular/upgrade/**/*.ts"],
        exclude = UPGRADE_STATIC_SRCS + [
            "rc/modules/@angular/upgrade/testing.ts",
            "rc/modules/@angular/upgrade/testing/**",
            "rc/modules/@angular/upgrade/test/**",
        ],
    ),
    deps = [
        ":angular2_es6_subset",
        ":compiler",
        ":core",
        ":platform_browser",
        ":platform_browser_dynamic",
        ":upgrade",
        "//third_party/javascript/node_modules/rxjs:canary",
    ],
)
```

(We don't let anyone new start using the "upgrade_dynamic" module)

**Comment by petebacondarwin** [comment]
Right, so if I split out the bits from `"rc/modules/@angular/upgrade/src/util.ts",` that depend upon the compiler into a new file (i.e. module) then it shouldn't break the static stuff.

**Comment by petebacondarwin** [comment]
@alexeagle and @vicb I have added a commit that extracts the JIT compiler dependent bits into the `upgrade_adapter.ts` file, so it should not cause problems with `upgrade/static` any more.

Please test and squash this new commit into the one with the similar commit message, if you are happy.

**Comment by youdz** [comment]
Thanks so much for getting this through even with all the extra work it needed to keep updating it to the latest structure and standards. 🎉

**Comment by petebacondarwin** [comment]
Sorry it took so long @youdz

**Comment by jensbodal** [comment]
For those like me that stumble across here, this is included in `4.0.0` but not `2.4.9`.

http://plnkr.co/edit/6SFy7nbK6TgtfQ44tMfs?p=preview

Change `ngVer` in `systemjs.config.js` to `4.0.0` to see it work.
